File: gek-core/src/main/java/xyz/fslabo/common/io/BytesInputStream.java
package xyz.fslabo.common.io;

import java.io.IOException;
import java.io.InputStream;

final class BytesInputStream extends InputStream {

    private final byte[] buf;
    private int pos;
    private int mark = -1;
    private final int count;

    BytesInputStream(byte[] buf) {
        this(buf, 0, buf.length);
    }

    BytesInputStream(byte[] buf, int offset, int length) {
        IOMisc.checkReadBounds(buf, offset, length);
        this.buf = buf;
        this.pos = offset;
        this.count = Math.min(offset + length, buf.length);
    }

    public int read() {
        return (pos < count) ? (buf[pos++] & 0xff) : -1;
    }

    public int read(byte[] b, int off, int len) {
        IOMisc.checkReadBounds(b, off, len);
        if (len <= 0) {
            return 0;
        }
        if (pos >= count) {
            return -1;
        }
        int avail = count - pos;
        avail = Math.min(len, avail);
        System.arraycopy(buf, pos, b, off, avail);
        pos += avail;
        return avail;
    }

    public long skip(long n) {
        if (n <= 0) {
            return 0;
        }
        int avail = count - pos;
        avail = (int) Math.min(n, avail);
        if (avail <= 0) {
            return 0;
        }
        pos += avail;
        return avail;
    }

    public int available() {
        return count - pos;
    }

    public boolean markSupported() {
        return true;
    }

    public void mark(int readAheadLimit) {
        mark = pos;
    }

    public void reset() throws IOException {
        if (mark < 0) {
            throw new IOException("Mark has not been set.");
        }
        pos = mark;
    }

    public void close() {
    }
}

File: boat-core/src/main/kotlin/xyz/srclab/common/io/BytesInputStream.kt
package xyz.srclab.common.io

import xyz.srclab.common.base.checkRangeInBounds
import xyz.srclab.common.base.toUnsignedInt
import java.io.InputStream

/**
 * Makes array as source of [InputStream].
 */
open class BytesInputStream(
    val source: ByteArray,
    private val offset: Int,
    private val length: Int
) : InputStream() {

    init {
        checkRangeInBounds(offset, offset + length, 0, source.size)
    }

    private var pos = offset
    private var mark = offset

    override fun read(): Int {
        if (pos >= offset + length) {
            return -1
        }
        val cur = source[pos]
        pos++
        return cur.toUnsignedInt()
    }

    override fun read(b: ByteArray): Int {
        return read(b, 0, b.size)
    }

    override fun read(b: ByteArray, off: Int, len: Int): Int {
        checkRangeInBounds(off, off + len, 0, b.size)
        if (pos >= offset + length) {
            return -1
        }
        val remaining = offset + length - pos
        return if (remaining <= len) {
            System.arraycopy(source, pos, b, off, remaining)
            remaining
        } else {
            System.arraycopy(source, pos, b, off, len)
            len
        }
    }

    override fun available(): Int {
        return offset + length - pos
    }

    override fun mark(readlimit: Int) {
        mark = pos
    }

    override fun reset() {
        pos = mark
    }

    override fun markSupported(): Boolean {
        return true
    }

    /**
     * Return copy of range of [source].
     */
    fun toByteArray(): ByteArray {
        return source.copyOfRange(offset, offset + length)
    }
}